pkg/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,391 @@
+	for _, svc := range services {
+		ipPool := svc.ObjectMeta.Annotations[antreaagenttypes.ServiceExternalIPPoolAnnotationKey]
+		if svc.Spec.Type != corev1.ServiceTypeLoadBalancer || ipPool == "" || len(svc.Status.LoadBalancer.Ingress) == 0 {
+			continue
+		}
+		ip := net.ParseIP(svc.Status.LoadBalancer.Ingress[0].IP)
+		allocation := externalippool.IPAllocation{
+			ObjectReference: v1.ObjectReference{
+				Name:      svc.Name,
+				Namespace: svc.Namespace,
+				Kind:      svc.Kind,
+			},
+			IPPoolName: ipPool,
+			IP:         ip,
+		}
+		previousIPAllocations = append(previousIPAllocations, allocation)
+	}
+	succeededAllocations := c.externalIPAllocator.RestoreIPAllocations(previousIPAllocations)
+	for _, alloc := range succeededAllocations {
+		name := apimachinerytypes.NamespacedName{
+			Namespace: alloc.ObjectReference.Namespace,
+			Name:      alloc.ObjectReference.Name,
+		}
+		c.setIPAllocation(name, alloc.IPPoolName, alloc.IP)
+		klog.InfoS("Restored external IP", "service", name, "ip", alloc.IP, "pool", alloc.IPPoolName)
+	}
+}
+
+func (c *ServiceExternalIPController) setIPAllocation(name apimachinerytypes.NamespacedName, ipPool string, ip net.IP) {
+	c.ipAllocationMutex.Lock()
+	defer c.ipAllocationMutex.Unlock()
+	c.ipAllocationMap[name] = &ipAllocation{
+		ip:     ip,
+		ipPool: ipPool,
+	}
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the workqueue.
+func (c *ServiceExternalIPController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *ServiceExternalIPController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+	if key, ok := obj.(apimachinerytypes.NamespacedName); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected NamespacedName in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncService(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing Service %s, requeuing. Error: %v", key, err)
+	}
+	return true
+}
+
+func (c *ServiceExternalIPController) releaseExternalIP(service apimachinerytypes.NamespacedName) error {
+	c.ipAllocationMutex.Lock()
+	defer c.ipAllocationMutex.Unlock()
+	allocation, exists := c.ipAllocationMap[service]
+	if exists {
+		if err := c.externalIPAllocator.ReleaseIP(allocation.ipPool, allocation.ip); err != nil {
+			if err == externalippool.ErrExternalIPPoolNotFound {
+				// Ignore the error since the external IP Pool could be deleted.
+				klog.Warningf("Failed to release IP %s because IP Pool %s does not exist", allocation.ip, allocation.ipPool)
+			} else {
+				klog.ErrorS(err, "Failed to release IP", "ip", allocation.ip, "pool", allocation.ipPool)
```

Comment:
```suggestion
				klog.ErrorS(err, "Failed to release external IP", "service", service, "ip", allocation.ip, "pool", allocation.ipPool)
```
So do other logs. To keep same log style as the following log and avoid confusion between ipPool used by Pods and externalIPPool.

Reply:
ok. updated.